codegen/smithy-go-codegen/src/main/java/software/amazon/smithy/go/codegen/integration/MiddlewareStackSnapshotTests.java

```diff
@@ -0,0 +1,164 @@
+                    if errors.Is(err, fs.ErrNotExist) {
+                        return snapshotOK{}
+                    }
+                    if err != nil {
+                        return err
+                    }
+                    defer f.Close()
+                    expected, err := io.ReadAll(f)
+                    if err != nil {
+                        return err
+                    }
+                    if actual := stack.String(); actual != string(expected) {
+                        return fmt.Errorf("%s != %s", expected, actual)
+                    }
+                    return snapshotOK{}
+                }
+                """,
+                MapUtils.of(
+                        "errors", SmithyGoDependency.ERRORS, "fmt", SmithyGoDependency.FMT,
+                        "fs", SmithyGoDependency.FS, "io", SmithyGoDependency.IO,
+                        "middleware", SmithyGoDependency.SMITHY_MIDDLEWARE, "os", SmithyGoDependency.OS
+                ));
+    }
+
+    private GoWriter.Writable snapshotUpdaters(Model model, ServiceShape service, SymbolProvider symbolProvider) {
+        return GoWriter.ChainWritable.of(
+                TopDownIndex.of(model).getContainedOperations(service).stream()
+                        .map(it -> testUpdateSnapshot(it, symbolProvider))
+                        .toList()
+        ).compose();
+    }
+
+    private GoWriter.Writable snapshotTests(Model model, ServiceShape service, SymbolProvider symbolProvider) {
+        return GoWriter.ChainWritable.of(
+                TopDownIndex.of(model).getContainedOperations(service).stream()
+                        .map(it -> testCheckSnapshot(it, symbolProvider))
+                        .toList()
+        ).compose();
+    }
+
+    private GoWriter.Writable testUpdateSnapshot(OperationShape operation, SymbolProvider symbolProvider) {
+        return goTemplate("""
+                func TestUpdateSnapshot_$operation:L(t $testingT:P) {
+                    svc := New(Options{})
+                    _, err := svc.$operation:L($contextBackground:T(), nil, func(o *Options) {
+                        o.APIOptions = append(o.APIOptions, func(stack $middlewareStack:P) error {
+                            return updateSnapshot(stack, $operation:S)
+                        })
+                    })
+                    if _, ok := err.(snapshotOK); !ok && err != nil {
+                        t.Fatal(err)
+                    }
+                }
+                """,
+                MapUtils.of(
+                        "testingT", GoStdlibTypes.Testing.T,
+                        "contextBackground", GoStdlibTypes.Context.Background,
+                        "middlewareStack", SmithyGoTypes.Middleware.Stack,
+                        "operation", symbolProvider.toSymbol(operation).getName()
+                ));
+    }
+
+    private GoWriter.Writable testCheckSnapshot(OperationShape operation, SymbolProvider symbolProvider) {
```

Comment:
One question, if a user want to test an op's stack snapshot, does he need to add `updateSnapshot` and `testSnapshot` into APIOptions in order?

Reply:
No, none of this is user-facing. The generated test code handles calling these functions for us.